File: models/Gpt-3.5-turbo/py/code/graph_algorithms/Hamiltonian_Flights.py

MOD = int(1e9) + 7

def count_routes(n, m, flights):
    count = 0
    visited = [False] * n

    def backtrack(city):
        nonlocal count
        if city == n:
            count += 1
            return
        visited[city] = True
        for next_city in range(n):
            if not visited[next_city] and flights[city][next_city]:
                backtrack(next_city)
        visited[city] = False

    flights = [[False] * n for _ in range(n)]
    for a, b in flights:
        flights[a - 1][b - 1] = True

    backtrack(0)
    return count % MOD

# Read input
n, m = map(int, input().split())
flights = []
for _ in range(m):
    a, b = map(int, input().split())
    flights.append((a, b))

# Count routes and print the result
print(count_routes(n, m, flights))

File: models/Gpt-3.5-turbo/cpp/code/problems/graph_algorithms/Hamiltonian_Flights.cpp

#include <iostream>
#include <vector>

using namespace std;

const int MOD = 1e9 + 7;

int n, m;
vector<vector<int>> flights;
vector<bool> visited;

int countRoutes(int city) {
    if (city == n) {
        // Reached Lehmälä, all cities visited once
        return 1;
    }

    int routes = 0;
    visited[city] = true;

    for (int i = 0; i < m; i++) {
        int from = flights[i][0];
        int to = flights[i][1];

        if (from == city && !visited[to]) {
            routes += countRoutes(to);
            routes %= MOD;
        }
    }

    visited[city] = false; // Reset visited for backtracking
    return routes;
}

int main() {
    cin >> n >> m;

    flights.resize(m, vector<int>(2));
    visited.resize(n + 1, false);

    for (int i = 0; i < m; i++) {
        cin >> flights[i][0] >> flights[i][1];
    }

    int routes = countRoutes(1); // Start from Syrjälä (city 1)
    cout << routes << endl;

    return 0;
}
